experimental/single_number_formatter/src/ule.rs

```diff
@@ -0,0 +1,112 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use zerovec::ule::{AsULE, ZeroVecError, ULE};
+
+use crate::provider::{CurrencyPatterns, PatternSelection};
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[repr(transparent)]
+pub struct CurrencyPatternsULE([u8; 3]);
+
+impl CurrencyPatternsULE {}
+
+// Safety (based on the safety checklist on the ULE trait):
+//  1. CurrencyPatternsULE does not include any uninitialized or padding bytes.
+//     (achieved by `#[repr(transparent)]` on a ULE type)
+//  2. CurrencyPatternsULE is aligned to 1 byte.
+//     (achieved by `#[repr(transparent)]` on a ULE type)
+//  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
+//  4. The impl of validate_byte_slice() returns an error if there are extra bytes.
+//  5. The other ULE methods use the default impl.
+//  6. CurrencyPatternsULE byte equality is semantic equality.
```

Comment:
Note: the two ignored bits can currently violate the "byte equality is semantic equality" non-safety invariant. It looks like in other custom impls such as `FieldLengthULE` and `CaseMapDataULE` we check for used bits in reserved zones, but personally I think it would be okay to GIGO on them in order for data to be more backwards compatible. CC @Manishearth

Reply:
the first and third bit are ignored

Reply:
They were ignored in the first version of the PR on these lines:

```rust
        let short_pattern_standard = if first_byte & 0b0100_0000 != 0 {
            PatternSelection::StandardAlphaNextToNumber
        } else {
            PatternSelection::Standard
        };

        let narrow_pattern_standard = if first_byte & 0b0001_0000 != 0 {
            PatternSelection::StandardAlphaNextToNumber
        } else {
            PatternSelection::Standard
        };
```

Anything could show up in bytes 1 and 3 and be ignored.

_However_, I am perfectly fine saying that `0b00, 0b10, 0b11` all map to Standard and `0b01` maps to StandardAlphaNextToNumber.

Reply:
do we have to use all the bits ?

Reply:
The problem is this:

```
let a_ule = CurrencyPatternsULE([0b0000_0000, 0b0000_0000, 0b0000_0000]);
let b_ule = CurrencyPatternsULE([0b1000_0000, 0b0000_0000, 0b0000_0000]);
let a = a.from_unaligned();
let b = b.from_unaligned();

assert_eq!(a_ule == b_ule, a == b);
```

This fails, and violates the sixth invariant that you listed above:
```
//  6. CurrencyPatternsULE byte equality is semantic equality.
```

Reply:
One way around this could be to add all four variants in CurrencyPatterns, naming the others `ReservedA` and `ReservedB`, for example.

Reply:
The other would be to return false in the validation function if the first and third bit are set.